ch_util/cal_utils.py

```diff
@@ -2161,3 +2166,315 @@ def _dec_to_el(dec):
+        d["valid"] = len(split_id) > 3
+
+        # If it's not a valid update we shouldn't try to extract everything else
+        if not d["valid"]:
+            continue
+
+        d["gen_time"] = ctime.datetime_to_unix(ctime.timestr_to_datetime(split_id[1]))
+        d["interpolated"] = "transition" in split_id
+        d["source_name"] = split_id[2].lower()
+
+        # Calculate the source transit time, and sanity check it
+        source = _source_dict[d["source_name"]]
+        d["source_transit"] = ephemeris.transit_times(
+            source, d["gen_time"] - 24 * 3600.0
+        )
+        cal_diff_hours = (d["gen_time"] - d["source_transit"]) / 3600
+        if cal_diff_hours > 3:
+            logger.warn(
+                f"Transit time ({ctime.unix_to_datetime(d['source_transit'])}) "
+                f"for source {d['source_name']} was a surprisingly long time "
+                f"before the gain update time ({cal_diff_hours} hours)."
+            )
+
+    # Array to store the extracted times in
+    reftime = np.zeros(len(collapsed_ids), dtype=np.float64)
+    reftime_prev = np.zeros(len(collapsed_ids), dtype=np.float64)
+    interp_start = np.zeros(len(collapsed_ids), dtype=np.float64)
+    interp_stop = np.zeros(len(collapsed_ids), dtype=np.float64)
+
+    # Iterate forward through the updates, setting transit times, and keeping track of
+    # the last valid update. This is used to set the previous source transit and the
+    # interpolation start time for all blended updates
+    last_valid_non_interpolated = None
+    for ii, state_id in enumerate(collapsed_ids):
+
+        valid_id = not np.ma.is_masked(state_id)
+        update = gain_info_dict[state_id] if valid_id else {}
+        valid = valid_id and update["valid"]
+
+        reftime[ii] = update["source_transit"] if valid else np.nan
+
+        if valid and update["interpolated"] and last_valid_non_interpolated is not None:
+            reftime_prev[ii] = reftime[last_valid_non_interpolated]
+            interp_start[ii] = times[last_valid_non_interpolated]
+        else:
+            reftime_prev[ii] = np.nan
+            interp_start[ii] = np.nan
+
+        if valid and not update["interpolated"]:
+            last_valid_non_interpolated = ii
+
+    # To identify the end of the interpolation periods we need to iterate
+    # backwards in time. As before we need to keep track of the last valid update
+    # we see, and then we set the interpolation end in the same manner.
+    last_valid_non_interpolated = None
+    for ii, state_id in list(enumerate(collapsed_ids))[::-1]:
+        valid_id = not np.ma.is_masked(state_id)
+        update = gain_info_dict[state_id] if valid_id else {}
+        valid = valid_id and update.get("valid", False)
+
+        if valid and update["interpolated"] and last_valid_non_interpolated is not None:
+            interp_stop[ii] = times[last_valid_non_interpolated]
```

Comment:
```suggestion
        if valid and update["interpolated"] and last_non_interpolated is not None:
            interp_stop[ii] = times[last_non_interpolated]
```
This would stop interpolating using the last non-interpolated update, whether or not it is valid.  This is mainly for consistency with `interp_start`.  Since the invalid updates are due to acquisition restarts, I don't think we can ever get into this state.